### ISSUE TITLE
Object extraction?

### ISSUE THREAD

**Issue by zachernuk** [immersive-web/model-element#102]

**Comment by m-blix** [comment]
What are the use cases you envision for the addressability?

One use case that could be useful is the ability to update a model after extraction. The benefit would be to have the user only have to position something accurately once, with model updates becoming much quicker.

1. User clicks on webpage to extract couch model A
2. User positions model at position X (potentially a somewhat lengthy operation to position just so)
3. User clicks on a webpage button for couch B
4. Couch B replaces couch A, at X


```html
<model id="couch">
  <source src="modern-couch.gs" type="model/splat">
</model>

<button id=couch-b-btn">Antique Couch</button>
```

```js
const modelEl = document.getElementById('couch');
const extractedModelRef = null;

const modelBButton = document.getElementById('couch-b-btn');
modelBButton.addEventListener('click', (e) => {
   if (extractedModelRef === null) return;
   extractedModelRef.src = 'antique-couch.gs';
});

modelEl.addEventListener('startextraction', (e) => {
   extractedModelRef = e.target; // some kind of reference to an extracted model
});

// user manually removes model from environment
modelEl.addEventListener('endextraction', (e) => {
  extractedModelRef = null;
});

```

**Comment by zachernuk** [comment]
That's right - the extracted model would remain subordinate to the page context but be able to exist at a position of a user's choosing, which doesn't need to be divulged to the page context, which would be a vector for some privacy risk.

It also affords the opportunity to apply a true scale to extracted content, given that the bounds of a page/panel may also be the subject of selective scaling, as is the case in visionOS.

**Comment by KooIaIa** [comment]
If the model is sub-orientated to the page context then it could follow the same 3D Transformation pattern that has been proposed for "Detached CSS". 

Like:   transform: rotate3d(1, 1, 1, 30deg)
    matrix3d(1, 0, 0, 0, 0, 1, 6, 0, 0, 0, 1, 0, 50, 100, 0, 1.1);

The same system that could place a <model> could also place a webpage <img> and eventually any 3D <html> element in WebXR.

**Comment by m-blix** [comment]
Do we need to discuss extraction itself or can we consider that up to the UA, just as img spec doesn't need to specify what a UA can do? (i.e. a UA can let user 'extract' an image via a context menu -> 'Save Image as..') I think maybe not.

In terms of the transform of the extracted model, I think there is an advantage to potentially not letting the page update the transform as in my example above the user would want the model position to not change. Perhaps changing the model transform could be conditional.

**Comment by cabanier** [comment]
> That's right - the extracted model would remain subordinate to the page context but be able to exist at a position of a user's choosing, which doesn't need to be divulged to the page context, which would be a vector for some privacy risk.

I'm unsure if that would be a good idea. Once extracted, I would expect that the model is no longer owned by the page.

**Comment by zachernuk** [comment]
I think the term 'extraction' can be misleading about the function of such a mode here, here's another try:

I think it's desirable to have an object that can be both associated with a page _and_ exist at full-scale. That way the model can be updated in response to page actions, e.g. to play/pause to different presentations of a reclining sofa, or different colors (for example, being stored in different frame-ranges of a model) 

There may be a difference between standard drag-and-drop that we'd want to address, but I'm very interested in how people would like to meet this general need on the spatial web.

**Comment by KooIaIa** [comment]
Can you think of more use-cases then a reclining sofa? A murphy bed? It stinks all the awesome uses for <model> are not being considered because of some privacy-risk. It doesn't feel like this would address general needs beyond these highly specific e-commerence ideas that WebXR is already used to solve. iOS only just recently got WebXR support in beta temporarily but if it was added sooner it would already be the de-facto way to address this.

**Comment by zachernuk** [comment]
(We didn't get to this last year, I'd like to discuss it next week)
/agenda 

I think this is a valuable capability to pursue anytime spatial content has a strong reason to be situated in space, and at a true (human-scale) size. Even with no other capabilities, that includes commerce, the heritage sector, data visualization and engineering applications.

While it's true that WebXR has the ability to meet these needs, it's through a totally separate track of work that cannot reuse other web capabilities directly, and the privacy risks you mention are significant enough that it's not necessarily _desirable_ for it to be the de-facto way to address them without a really compelling reason.

Finally, `<model>` as it is proposed today isn't the end goal. As always, the challenge with spatial content is how to bring progressively more spatial capability _into_ the existing taxonomy of web without falling off a cliff of irreducible complexity.

**Comment by zachernuk** [comment]
While it's not the exact same situation, the [Picture-in-picture (PiP) API](https://developer.mozilla.org/en-US/docs/Web/API/Picture-in-Picture_API) covers similar functionality:
*  A PiP element is _normally_ a video, but it can also be some hierarchy of DOM content, such that people can provide custom controls or other information. 
* The user gets to position a PiP element in their OS, but the invoking context doesn't know where it is.
* The PiP element is destroyed if the parent context is lost, e.g. from closing the tab or navigating away.

**Comment by KooIaIa** [comment]
> I think this is a valuable capability to pursue anytime spatial content has a strong reason to be situated in space

So the idea is anytime someone wants to view a spatial visualization they need to drag a model out of the page or click a PiP-like button that will randomly place it? And then there is no web styling, instead its all USD animations instead of CSS? USD will essentially become an invaluable trillion dollar format for engineering, e-commerce, data visualization, and heritage usable across all browsers? 

> While it's true that WebXR has the ability to meet these needs, it's through a totally separate track of work that cannot reuse other web capabilities directly

Re-usable web capability in WebXR should be a priority but year after year isn't. HTML / Web Capabilities in WebXR would address this.

>spatial capability _into_ the existing taxonomy of web without falling off a cliff of irreducible complexity.

Wanting USD support in every browser is a massive cliff of irreducible complexity. It's an incredibly complex format your company is deeply invested in and you haven't even been open to other \<sources\>. Enhancing the capabilities of WebXR for web visualizations avoids this cliff completely.

**Comment by zachernuk** [comment]
> [PiP-like activity] will randomly place [web content]?

It's not random - PiP is a capability that allows the user to place content on all supported platforms today, it would be reasonable to aim to do the same in the future.

>  all USD animations instead of CSS?

The explainer indicates that accessing a scene graph and/or the ability to compose DOM content along with `<model>` is a desirable goal - the critical question for any development roadmap is what we need to build _first._ 

It's my belief that `<model>` is the best first step for all of this web-friendly composition, and allows web authors to push on a different set of boundaries than WebXR development.

> Format discussion

I believe the issue at hand can be discussed independently to questions of format.

> WebXR avoiding cliffs

Unfortunately, integrating web content (or allowing WebXR content to be remain active alongside other experiences) _does_ hit very challenging security boundaries that pose similarly existential risks for the technology.

**Comment by cabanier** [comment]
> While it's not the exact same situation, the [Picture-in-picture (PiP) API](https://developer.mozilla.org/en-US/docs/Web/API/Picture-in-Picture_API) covers similar functionality

Interesting! What happens when you use this feature on AVP? FWIWI Quest browser doesn't support it.
